### ISSUE TITLE
TimeoutPersisterReceiver in V5: New timeouts trigger querying the store two times

### ISSUE THREAD

**Issue by bording** [Particular/NServiceBus#3833]
The issue described in #3300 and fixed in #3540 for V6 is a problem for the V5 `TimeoutPersisterReceiver` as well. 

@indualagarsamy and I have been working on a support case where this is the root cause of the problems being encountered.

Seeing how this needlessly creates duplicate messages, I propose we backport the fix to V5 and release a hotfix.

Any objections?

**Comment by indualagarsamy** [comment]
@Particular/nservicebus-maintainers - Please respond on if we can back port this fix. I agree with @bording, this needs to be back ported IMO.

**Comment by timbussmann** [comment]
quoting from the linked issue (#3300):

> although we query again immediately, we shouldn't receive the same timeout again, since we adjusted startSlice on the first query and the persistence should exclude timeouts which exactly equal startSlice

there was an nhibernate persistence issue which didn't  handle the `startSlice` value properly (it should exclude timeouts which are **equal** or earlier than `startslice`) if I remember correctly. Can you define the scenario where an immediate second query to the timeout would return the same timeout again?

**Comment by bording** [comment]
@timbussmann I think the issue you're referring to is https://github.com/Particular/NServiceBus.NHibernate/issues/145, but no change has actually been made related to that yet.

This is how it's currently defined in all of the persistences I've looked at while investigating this:

https://github.com/Particular/NServiceBus.NHibernate/blob/develop/src/NServiceBus.NHibernate/TimeoutPersisters/TimeoutPersister.cs#L44

The second polling does find the same timeout if the Dispatcher satellite hasn't yet had time to process the dispatch control message, and this is the behavior being seen in the support case.

**Comment by bording** [comment]
So, after talking with @timbussmann a bit about this, it looks like we've got two things working together to cause this issue.

First, as pointed out, we are triggering an immediate re-poll without a delay in some cases. On it's own, this isn't a problem, though I'm not clear on why this was considered desirable behavior (or is it a bug?)

The other part of this that caused the problem is how the persistences implement `GetNextChunk`, specifically in how they use `startSlice`.

Looking at all of the persistence implementations (both official and community) this value is being used in two different ways. A couple are using `Time > startSlice`, but most are using `Time >= startSlice`

Given that the V5 poller can do an immediate re-poll, using `Time >= startSlice` means the same timeout can be included in both the original query and the re-poll query.

Removing the immediate re-poll does mostly help mitigate this, but it doesn't completely prevent the problem, since the next retry attempt value could be so small that it skips over the Thread.Sleep.

II believe that updating all of the persistences to use `>` instead of `>=` should make this not happen regardless of the immediate re-poll logic happening or not.

Should we look into opening PRs to make this change where needed instead of making a core change, or should we do both?

**Comment by timbussmann** [comment]
as far as I know `Time > startSlice` is definitely the one we should use. Wondering why this was changed for raven 4.0 (seems to be a mistake, but not sure about the impact because of the removed immediate repoll).

Since there is no 100% way to avoid duplicates outside of DTC scenarios I'm a bit on the fence about the effort to go down the "remove the immediate repoll" rabbit hole seeing as we fixed this in v6. Definitely :+1: to raise PRs against the transports and to verify whether this solves the pain experienced by the affected users. If they still experience an unusual amount of duplicates, let's remove the repolling mechanism.
